### DISCUSSION TITLE
Reconnect to Terminal after window reload

### DISCUSSION THREAD

**Discussion by DavyLandman** [microsoft/vscode-discussions#909] [Extension Development QnA] [answered]
We have an terminal that is started using `createTerminal` and we have a side-channel between the extension and the terminal (localhost random port that the extension open).

If the window is reloaded, VS Code nicely keeps the terminal process around, but which events can we use to detect such a case? And where should we keep some data transient? is the `workspaceState` the right place to store some "temporary" information like port information?

**Comment by isidorn** [comment]
@tyriar might be able to help here

**Comment by DavyLandman** [comment]
Just to help out @Tyriar when he get's around: we start a process using createTerminal's shellPath&shellArgs (see [source](https://github.com/usethesource/rascal-language-servers/blob/edf9f36aba0323305f9433134654fe43e68bb228/rascal-vscode-extension/src/RascalExtension.ts#L134-L140)). Before that happens our extension opens a random tcp port onl localhost, and passes that one into the terminal as an argument. They use this to communicate a few things in a cross platform way. If VS Code is reloaded/restarted, I would like to try and reconnect the terminal to the current extension, but I do not know where to keep state and how to recover.

**Comment by Tyriar** [comment, marked as answer]
The PowerShell extension has a similar problem, I think what they end up doing is setting `isTransient` to true and recreating the terminal based on the previous state. You could also potentially use the `processId` of the `Terminal` to reconnect to it?

`workspaceState` should be fine, if it ends up being a lot of data you may want to persist to a file.

**Reply by DavyLandman** [reply]
okay, hmm, damn. Sending messages to the processId will be messy as well. But I'll give that a go.

The `workspaceState` is not cleared after a VS Code restart right? So I have to be sure to detect that difference. But thanks, I have some places to look at.

**Reply by Tyriar** [reply]
Yep `workspaceState` is retained across restarts/reloads
